File: backend-springboot/src/main/java/com/monitor/controller/ErrorController.java
package com.monitor.controller;

import com.monitor.entity.ErrorLog;
import com.monitor.service.ErrorService;
import com.monitor.service.ElasticsearchService;
import com.monitor.util.DateUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * 错误监控控制器
 */
@RestController
@RequestMapping("/error")
@RequiredArgsConstructor
public class ErrorController {
    
    private final ErrorService errorService;
    private final ElasticsearchService elasticsearchService;
    
    @Value("${default-project-id:001}")
    private String defaultProjectId;
    
    @PostMapping("/report")
    public ResponseEntity<Map<String, Object>> report(@RequestBody Map<String, Object> body) {
        return handleReport(body);
    }
    
    @GetMapping("/report")
    public ResponseEntity<Map<String, Object>> reportGet(@RequestParam String data) {
        try {
            // URL 解码并解析 JSON
            String decoded = java.net.URLDecoder.decode(data, java.nio.charset.StandardCharsets.UTF_8);
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            Map<String, Object> body = mapper.readValue(decoded, Map.class);
            return handleReport(body);
        } catch (Exception e) {
            System.err.println("Error report GET failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "数据解析失败"));
        }
    }
    
    private ResponseEntity<Map<String, Object>> handleReport(Map<String, Object> body) {
        try {
            // 调试日志：记录接收到的错误信息
            if ("resource".equals(body.get("type"))) {
                System.out.println("📦 Resource error received: " + Map.of(
                    "type", body.getOrDefault("type", ""),
                    "message", body.getOrDefault("message", ""),
                    "url", body.getOrDefault("url", ""),
                    "userId", body.getOrDefault("userId", "")
                ));
            }
            
            ErrorLog error = errorService.report(
                defaultProjectId,
                (String) body.getOrDefault("type", "js"),
                (String) body.getOrDefault("message", "Unknown error"),
                (String) body.get("stack"),
                (String) body.get("url"),
                body.get("line") != null ? ((Number) body.get("line")).intValue() : null,
                body.get("col") != null ? ((Number) body.get("col")).intValue() : null,
                (String) body.get("userAgent"),
                (String) body.get("userId"),
                (String) body.get("sessionId")
            );
            
            if ("resource".equals(body.get("type"))) {
                System.out.println("✅ Resource error saved: " + error.getId());
            }
            
            // 异步写入 Elasticsearch（不阻塞主流程）
            Map<String, Object> logData = new HashMap<>();
            logData.put("projectId", defaultProjectId);
            logData.put("type", "error");
            logData.put("userId", body.get("userId"));
            logData.put("sessionId", body.get("sessionId"));
            logData.put("url", body.get("url"));
            logData.put("timestamp", new java.util.Date().toInstant().toString());
            logData.put("errorType", body.getOrDefault("type", "js"));
            logData.put("errorMessage", body.getOrDefault("message", "Unknown error"));
            logData.put("stack", body.get("stack"));
            logData.put("line", body.get("line"));
            logData.put("col", body.get("col"));
            logData.put("userAgent", body.get("userAgent"));
            logData.put("message", body.getOrDefault("message", "Unknown error"));
            
            try {
                elasticsearchService.writeLog(logData);
            } catch (Exception e) {
                System.out.println("⚠️ Elasticsearch write failed (non-blocking): " + e.getMessage());
            }
            
            return ResponseEntity.ok(Map.of("success", true, "id", error.getId()));
        } catch (Exception e) {
            System.err.println("Error report failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "上报失败"));
        }
    }
    
    @GetMapping("/list")
    public ResponseEntity<Map<String, Object>> list(
            @RequestParam(required = false) String type,
            @RequestParam(required = false, defaultValue = "1") Integer page,
            @RequestParam(required = false, defaultValue = "20") Integer pageSize,
            @RequestParam(required = false) String startTime,
            @RequestParam(required = false) String endTime) {
        
        Map<String, Object> result = errorService.findList(defaultProjectId, type, page, pageSize, startTime, endTime);
        
        // 格式化日期字段
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> data = (List<Map<String, Object>>) result.get("data");
        if (data != null) {
            data = data.stream().map(item -> {
                Map<String, Object> formatted = new HashMap<>(item);
                if (item.get("timestamp") != null) {
                    formatted.put("timestamp", DateUtil.formatDateTime(item.get("timestamp")));
                }
                if (item.get("firstSeen") != null) {
                    formatted.put("firstSeen", DateUtil.formatDateTime(item.get("firstSeen")));
                }
                if (item.get("lastSeen") != null) {
                    formatted.put("lastSeen", DateUtil.formatDateTime(item.get("lastSeen")));
                }
                return formatted;
            }).collect(Collectors.toList());
        }
        
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", data != null ? data : List.of(),
            "total", result.get("total"),
            "page", result.get("page"),
            "pageSize", result.get("pageSize")
        ));
    }
    
    @GetMapping("/detail/{id}")
    public ResponseEntity<Map<String, Object>> detail(@PathVariable String id) {
        Optional<ErrorLog> errorOpt = errorService.findById(id);
        if (errorOpt.isEmpty()) {
            return ResponseEntity.ok(Map.of("success", false, "message", "错误不存在"));
        }
        
        ErrorLog error = errorOpt.get();
        Map<String, Object> errorData = new HashMap<>();
        errorData.put("id", error.getId());
        errorData.put("type", error.getType());
        errorData.put("message", error.getMessage());
        errorData.put("stack", error.getStack());
        errorData.put("url", error.getUrl());
        errorData.put("line", error.getLine());
        errorData.put("col", error.getCol());
        errorData.put("userAgent", error.getUserAgent());
        errorData.put("userId", error.getUserId());
        errorData.put("sessionId", error.getSessionId());
        errorData.put("timestamp", DateUtil.formatDateTime(error.getTimestamp()));
        errorData.put("firstSeen", DateUtil.formatDateTime(error.getFirstSeen()));
        errorData.put("lastSeen", DateUtil.formatDateTime(error.getLastSeen()));
        errorData.put("count", error.getCount());
        errorData.put("affectedUsers", error.getAffectedUsers());
        errorData.put("context", error.getContext());
        errorData.put("platform", error.getPlatform());
        errorData.put("appName", error.getAppName());
        errorData.put("isOnline", error.getIsOnline());
        errorData.put("isPWA", error.getIsPWA());
        
        return ResponseEntity.ok(Map.of("success", true, "data", errorData));
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> stats() {
        Map<String, Object> stats = errorService.getStats(defaultProjectId);
        return ResponseEntity.ok(Map.of("success", true, "data", stats));
    }
}

File: backend-midway/src/controller/error.controller.ts
import { Controller, Get, Post, Inject, Body, Param, Query, Config } from '@midwayjs/core';
import { ApiTags, ApiOperation, ApiParam, ApiQuery, ApiBody } from '@midwayjs/swagger';
import { ErrorService } from '../service/error.service';
import { ElasticsearchService } from '../service/elasticsearch.service';
import { formatDateTime } from '../utils/date.util';

@Controller('/error')
@ApiTags(['error'])
export class ErrorController {
  @Inject()
  errorService: ErrorService;

  @Inject()
  elasticsearchService: ElasticsearchService;

  @Config('defaultProjectId')
  defaultProjectId: string;

  @Post('/report')
  @ApiOperation({ summary: '上报错误数据' })
  @ApiBody({ description: '错误信息' })
  async report(@Body() body: any) {
    return this.handleReport(body);
  }

  @Get('/report')
  @ApiOperation({ summary: '上报错误数据（GET方式，用于img上报）' })
  @ApiQuery({ name: 'data', required: true, description: 'JSON编码的错误信息' })
  async reportGet(@Query('data') dataStr: string) {
    try {
      const body = JSON.parse(decodeURIComponent(dataStr));
      return this.handleReport(body);
    } catch (error) {
      console.error('Error report GET failed:', error);
      return { success: false, message: '数据解析失败' };
    }
  }

  private async handleReport(body: any) {
    try {
      // 调试日志：记录接收到的错误信息
      if (body.type === 'resource') {
        console.log('📦 Resource error received:', {
          type: body.type,
          message: body.message,
          url: body.url,
          userId: body.userId
        });
      }
      
      const error = await this.errorService.report({
        projectId: this.defaultProjectId,
        type: body.type || 'js',
        message: body.message || 'Unknown error',
        stack: body.stack,
        url: body.url,
        line: body.line,
        col: body.col,
        userAgent: body.userAgent,
        userId: body.userId,
        sessionId: body.sessionId,
      });

      if (body.type === 'resource') {
        console.log('✅ Resource error saved:', error._id);
      }

      // 异步写入 Elasticsearch（不阻塞主流程）
      this.elasticsearchService.writeLog({
        projectId: this.defaultProjectId,
        type: 'error',
        userId: body.userId,
        sessionId: body.sessionId,
        url: body.url,
        timestamp: new Date().toISOString(),
        errorType: body.type || 'js',
        errorMessage: body.message || 'Unknown error',
        stack: body.stack,
        line: body.line,
        col: body.col,
        userAgent: body.userAgent,
        message: body.message || 'Unknown error',
      }).catch((err) => {
        // Elasticsearch 写入失败不影响主流程，只记录错误
        console.warn('⚠️ Elasticsearch write failed (non-blocking):', err.message);
      });

      return { success: true, id: error._id };
    } catch (error) {
      console.error('Error report failed:', error);
      return { success: false, message: '上报失败' };
    }
  }

  @Get('/list')
  @ApiOperation({ summary: '获取错误列表' })
  @ApiQuery({ name: 'type', required: false, description: '错误类型' })
  @ApiQuery({ name: 'page', required: false, description: '页码' })
  @ApiQuery({ name: 'pageSize', required: false, description: '每页条数' })
  async list(@Query() query: any) {
    const { type, page, pageSize, startTime, endTime } = query;

    const result = await this.errorService.findList({
      projectId: this.defaultProjectId,
      type,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
      startTime,
      endTime,
    });

    return {
      success: true,
      data: result.data.map(e => {
        const errorData = (e as any).toObject ? (e as any).toObject() : e;
        return {
          id: errorData._id?.toString() || errorData.id,
          type: errorData.type,
          message: errorData.message,
          url: errorData.url,
          userId: errorData.userId || 'anonymous',
          timestamp: formatDateTime(errorData.timestamp),
          count: errorData.count,
          firstSeen: formatDateTime(errorData.firstSeen),
          lastSeen: formatDateTime(errorData.lastSeen),
          affectedUsers: errorData.affectedUsers,
        };
      }),
      total: result.total,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
    };
  }

  @Get('/detail/:id')
  @ApiOperation({ summary: '获取错误详情' })
  @ApiParam({ name: 'id', description: '错误ID' })
  async detail(@Param('id') id: string) {
    const error = await this.errorService.findById(id);
    if (!error) {
      return { success: false, message: '错误不存在' };
    }
    // 将 MongoDB 文档转换为普通对象，确保所有字段都被包含
    // Typegoose 返回的文档可能有 toObject 方法，但类型定义中可能没有
    const errorData = (error as any).toObject ? (error as any).toObject() : error;
    return { 
      success: true, 
      data: {
        id: errorData._id?.toString() || errorData.id,
        type: errorData.type,
        message: errorData.message,
        stack: errorData.stack,
        url: errorData.url,
        line: errorData.line,
        col: errorData.col,
        userAgent: errorData.userAgent,
        userId: errorData.userId,
        sessionId: errorData.sessionId,
        timestamp: formatDateTime(errorData.timestamp),
        firstSeen: formatDateTime(errorData.firstSeen),
        lastSeen: formatDateTime(errorData.lastSeen),
        count: errorData.count,
        affectedUsers: errorData.affectedUsers,
        context: errorData.context,
        platform: errorData.platform,
        appName: errorData.appName,
        isOnline: errorData.isOnline,
        isPWA: errorData.isPWA,
      }
    };
  }

  @Get('/stats')
  @ApiOperation({ summary: '获取错误统计' })
  async stats() {
    const stats = await this.errorService.getStats(this.defaultProjectId);
    return { success: true, data: stats };
  }
}

